### PR TITLE
contact top story and error handling

### PR DESCRIPTION
**Pull request by MAlngaawy**
This pr isn't for review or approve 
 I trying to implement storybook at simple component ( ContactTop )
to show to Ayoub if it's okay

### CODE REVIEW

File: src/components/Header/index.jsx [hunk 1]
```diff
@@ -1,12 +1,13 @@
 import React from 'react';
-import ContactTop from '../ContactTop';
+// import ContactTop from '../ContactTop';
+import { ContactTopContainer } from '../../otherPagesComponent/ContactTop';
```
Comment (Nilegfx) [inline src/components/Header/index.jsx:3]: there is no need to create `otherPagesComponent` you should create them directly in `components`

File: src/stories/3-ContactTop.stories.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  decorators: [withKnobs]
+};
+
+export const CotactTopComponent = () => {
+  return (
+    <ContactTop
+      socialIconsList={data}
+      loading={boolean('loading', false)}
+      error={boolean('error', false)}
+      clicked={boolean('clicked', false)}
```
Comment (Nilegfx) [inline src/stories/3-ContactTop.stories.js:18]: when you want to test event handlers like this one `clicked`, you should use storybook's [actions addon](https://github.com/storybookjs/storybook/tree/master/addons/actions)

```js
import { action } from '@storybook/addon-actions';
<ContactTop
      socialIconsList={data}
      loading={boolean('loading', false)}
      error={boolean('error', false)}
      clicked={action('Yellow-button-clicked')} //like this
/>
```

once you click on the button from within the story, you should see something like this in the actions panel
image

and in from the component implementation point of view, you should keep everything as it is like this:

```js<button
        onClick={clicked}
        className="btn w-2/12 h-full text-c100 text-sm font-bold bg-c300"
      >
        Start Donation
      </button>```

Comment (MAlngaawy) [reply src/stories/3-ContactTop.stories.js:18]: Thanks a lot i will  reconsider that
i would use action but i have confusing about the benefit of this .. i thought that I'm going to see the changes i wanted in the component ( like make section's background yellow )
then when i didn't get it .. i ignore using it
but now i know that this is just to tell me that this specific button has been clicked
Thanks again 👍 💯 
@Nilegfx

Comment (Nilegfx) [reply src/stories/3-ContactTop.stories.js:18]: think about it, if you want to make the background to be changed upon clicking the button, do you need to send a function from the outside into the component?

The reason why we send functions from outside is to invoke it to do things in a higher/another scope, the reason why storybook actions add-on exists is to tell you that your function was called and it shows you with which parameters it was called.

**an example**

if you changed the implementation of the component to be like: 
```js
//...
const ContactTop = ({ socialIconsList, loading, error, clicked }) => {

  // notice 1
  const yellowButtonClickHandler = () => {
    clicked(5);
  };

//...

<button
onClick={yellowButtonClickHandler} //notice 2
className="btn w-2/12 h-full text-c100 text-sm font-bold bg-c300"
>
Start Donation
</button>
}
```

and changed the container implementation to be like: 
```js
const ContactTopContainer = () => {
  const { data, loading, dataError } = useCharityAPI('/socialmedias');
  const [yellowButtonClickCounter, setYellowButtonClickCounter] = useState(0); //notice 3

  //notice 4
  const incrementYellowButtonCounterBy = number => {
    setYellowButtonClickCounter(yellowButtonClickCounter + number);
  };

  return (
    <ContactTop
      socialIconsList={data}
      loading={loading}
      error={dataError}
      clicked={incrementYellowButtonCounterBy} //notice 5
    />
  );
};
```

now in the story, when you click the button, the action panel will show you:
image


This is the benefit of it, now you know that your component called the function which it received as prop, which eventually passes it back to the parent/container component with the correct desired parameter

File: src/stories/contactTopData.json [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  {
+    "_id": "5ea5751af66232391cd860a7",
+    "fontawesome_icons": "fab fa-instagram",
+    "url": "www.instgram.com",
+    "createdAt": "2020-04-26T11:48:42.473Z",
+    "updatedAt": "2020-04-26T11:48:42.473Z",
+    "__v": 0,
+    "id": "5ea5751af66232391cd860a7"
+  }
+]
```
Comment (Nilegfx) [inline src/stories/contactTopData.json:38]: I do know that I did this in my live/demo, but as usual, please do not take what I do as standard, when I am on live, I sacrifice some best practices for the sake of time-saving. 

in this kind of mocked data, it is always better to keep it at the minimal, just the keys which you use in your component, the reset you should delete. 

By keeping only the used keys, you give the reader/others an idea which keys are being used by the desired component without them having to look inside and see the implementation. so in this example you use `id`, `url` and `fontawesome_icons` keys only. your mock might look simpler if it is something like: 

```js
[
  {
    "fontawesome_icons": "fab fa-facebook-square",
    "url": "www.facebook.com",
    "id": "5ea574f0f66232391cd860a4"
  },
  {
    "fontawesome_icons": "fab fa-twitter",
    "url": "www.twitter.com",
    "id": "5ea574faf66232391cd860a5"
  },
  {
    "fontawesome_icons": "fab fa-pinterest-p",
    "url": "www.pinterest.com",
    "id": "5ea57503f66232391cd860a6"
  },
  {
    "fontawesome_icons": "fab fa-instagram",
    "url": "www.instgram.com",
    "id": "5ea5751af66232391cd860a7"
  }
]
```

Comment (MAlngaawy) [reply src/stories/contactTopData.json:38]: thanks . i will fix it

**Comment by Nilegfx** [comment]
BTW, storybook is the whole library, what you write is just story/stories. so a more descriptive title for your PR could be **test contact top story**

File: src/components/ContactTop/index.jsx [hunk 4]
```diff
@@ -76,9 +107,26 @@ const ContactTop = () => {
...
+            className={cn(
+              'btn',
+              'w-2/12',
+              'h-full',
+              'text-c100',
+              'text-sm',
+              'font-bold',
+              'bg-c300',
+              {
+                hidden: !haveBtn
```
Comment (Nilegfx) [inline src/components/ContactTop/index.jsx:121]: Super Nice!
However, given that we have react now, lets do things better than hiding elements with css 😉

React is a tree of components renders multiple time, the less components/html to render the better, so you can easily now use `haveBtn` prop to conditionally render (add)/remove components.

Bounce tip:
You can also have a look at Higher-Order Components  which will help you to use one base component to create two different components (one with haveBtn:true and other without)
`TopContact' and `TopContactWithBtn'

File: src/stories/3-ContactTop.stories.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  decorators: [withKnobs]
+};
+
+export const CotactTopComponent = () => {
+  return (
+    <ContactTop
+      data={data}
+      loading={boolean('loading', false)}
+      error={boolean('error', false)}
+      haveBtn={boolean('haveButton', false)} // If Your Component Have a Donate Button
```
Comment (Nilegfx) [inline src/stories/3-ContactTop.stories.js:19]: Nice awe!

**Comment by MAlngaawy** [comment]
I have implemented 3 things in this PR 
1_ Contact Top Story
2_ Make Component Container
3_ implement Error Handling
You Can Look to it as a reference while you work ( If Ayoub And Ahmed Samier Accept what I did )

File: src/components/ContactTop/style.scss [hunk 6]
```diff
@@ -1,5 +1,6 @@
 .contact-top {
   color: rgba(255, 255, 255, 0.7);
+  height: 48px;
```
Comment (Tramsi) [inline src/components/ContactTop/style.scss:3]: Why have you added this height?

**Pull request merged**